src/OpenSearch/Transport.php

```diff
@@ -144,19 +144,19 @@ function ($response) {
      *
      * @return callable|array
      */
-    public function resultOrFuture(FutureArrayInterface $result, array $options = [])
+    public function resultOrFuture(Promise $result, array $options = [])
     {
         $response = null;
         $async = isset($options['client']['future']) ? $options['client']['future'] : null;
         if (is_null($async) || $async === false) {
             do {
                 $result = $result->wait();
-            } while ($result instanceof FutureArrayInterface);
+            } while ($result instanceof Promise);
         }
         return $result;
```

Comment:
Changing this return type from `FutureArrayInterface|array` (`callable|array` was wrong) to `Promise|array` will break BC for projects using the `future` client option to run in async mode as it changes the return type for them.

However, as the generated code does not give access to those options to the projects using them (generated endpoints class are the ones able to configure options), it seems like `Client::request()` is the only method actually allowing to actually run async. So maybe all the async support is a bunch of dead code.